### ISSUE TITLE
Blazor InputSelect sets wrong selected Option

### ISSUE THREAD

**Issue by zoran-petrovic-87** [dotnet/aspnetcore#41222]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Let's say that we have a really simple class with 2 properties, one is an integer and the other one is a string. I want to use EditForm with 2 InputSelect elements that are bind to these properties. The code looks like this:

```
@page "/test"
@using Microsoft.AspNetCore.Components
@using Newtonsoft.Json

<PageTitle>Test</PageTitle>
<h3>Test form</h3>

<EditForm Model="@_dummy" OnValidSubmit="@TestAsync">
    <DataAnnotationsValidator/>
    <ValidationSummary/>

    <div class="mb-3">
        <label for="string" class="form-label">String</label>
        <InputSelect id="string" @bind-Value="_dummy.StringProp" class="form-select">
            <option value="A">A</option>
            <option value="B">B</option>
            <option value="C">C</option>
        </InputSelect>
    </div>

    <div class="mb-3">
        <label for="int" class="form-label">Integer</label>
        <InputSelect id="int" @bind-Value="_dummy.IntProp" class="form-select">
            <option value="1">1</option>
            <option value="2">2</option>
            <option value="3">3</option>
        </InputSelect>
    </div>

    <button type="submit" class="btn btn-primary">Start</button>
</EditForm>

@code {
    readonly Dummy _dummy = new();

    private void TestAsync()
    {
        Console.WriteLine("Object state: " + JsonConvert.SerializeObject(_dummy));
    }

    class Dummy
    {
        public string? StringProp { get; set; }
        public int IntProp { get; set; }
    }

}
```

Here is how it looks like in a web browser:

![image](https://user-images.githubusercontent.com/6372824/163680842-698615f1-e3e2-4867-89da-330844f01181.png)

As we can see, if property is nullable, then Blazor will set the first option as selected. When we submit this form, the actual value of that property will not be the value of the selected option, it will be null.

### Expected Behavior

When we bind InputSelect to nullable property, it should not select any option by default. The behavior should be the same as with non-nullable properties

### .NET Version

6.0

**Comment by jessegood** [comment]
In your second example, the reason nothing is displayed is because `0` is not an available option.
The moment you add `0` as an option, then `0` will be displayed.
`<option value="0">0</option>`

So in order for Blazor to show nothing when you specify `null`, it would have to choose a value that is not one of the available options. For example, it could use the empty string `""`, but then what if the user adds the following as an option?
`<option value="">---Select something---</option>`

Now we are back to square one, so that is why I think there isn't a sensible value for this.

**Comment by zoran-petrovic-87** [comment]
@jessegood I think it is safer just to keep nothing pre-selected when property has _null_ as a default value.
It would actually work perfectly fine with the current implementation, since if a user doesn't interact with InputSelect then the property will keep it's original value. 

What causes bugs in production is - form shows "A" as selected option. User thinks that is OK and submit the data. But "A" is not submitted, value of a property is still _null_!

What should happen is - nothing is pre-selected (just as in integer case). If a user doesn't want to select that option we would get a _null_ (or any default) value. If user selects something then we get the selected value.

Basically we get a false representation of binded object on UI + we have inconsistent behavior depending on if property is nullable or not.

**Comment by javiercn** [comment]
@zoran-petrovic-87 I think this is expected.

What's happening here is that the bound default value `(null)` doesn't affect the underlying input select in any way, while the default value for an integer does.

As a result, that's why you see the difference between the first option being selected in one case and no option being selected in the second case. However, if you were to submit the form, in both cases you would get `null` in the first case and `0` in the second case, which doesn't match any of the options.

I think it is best/recommended to initialize the properties you bound to an input select with non-null values (empty string) to avoid this issue.

**Comment by zoran-petrovic-87** [comment]
Hi @javiercn I understand what is happening, but I still think that is wrong.
For example, if you create a form like this and ask users that they need to submit it with a value "A", I am pretty sure that 99% of them will not click on InputSelect since "A" is already selected. And on the backend we will get null. But if "A" was not preselected then we would get correct result on the backend.

**Comment by javiercn** [comment]
@zoran-petrovic-87 I'm not implying that its not a bug. I've added more info to capture some more details so that when we discuss the issue as part of our planning process we can have a more informed debate around it.

You bring up good points, however, I want to get some inputs from other folks on the team about it.

**Comment by MilesFan** [comment]
@zoran-petrovic-87  I think this is more of an HTML thing than a Blazor one.

Adding a **required** attriute and an empty option to the control has been working for me.

The empty option matches any unexpected value, and the required attribute prevents submitting with the null value in the empty option.

```
...
<InputSelect id="string" @bind-Value="_dummy.StringProp" class="form-select" required>
    <option></option>
    <option value="A">A</option>
    <option value="B">B</option>
    <option value="C">C</option>
</InputSelect>
...
```

Or use `<option>Please select</option>` to be more user-freindly.

**Comment by vgallegob** [comment]
I also had this problem, it was very annoying

**Comment by friedice5467** [comment]
Workaround I found was to just have an empty value and st style display to none 

`<option value="" style="display:none;"></option>`

**Comment by UniMichael** [comment]
Note that some browsers don't support `style="display:none;`. If you're trying to hide the option, you'll want to do `<option hidden disabled value=""></option>` instead (some browsers don't support `hidden` properly, so `disabled` will at least prevent the option from being selected).

**Comment by Kissaki** [comment]
A HTML select defaults to

1. the `selected` option
2. the first option

When you set the select value via JavaScript to a value that does not exist, it clears the rendered select content (dropdown does not have the option, but the select).  
When you set the select value to null, the select value is an empty string.

*(Tested in Firefox)*

---

Given that the HTML select covers a non-match, it is appropriate to make use of it for non-matches.

As per the original example, this works for int (non-nullable struct types).

Why would null not be handled the same way? Null is a non-match of the select options. It should render like a non-match of an int select. No option selected, and the value being the value.

The Blazor InputSelect with bound type param should transparently translate/encapsulate this behavior when the HTML select can not cover null as a value itself, vs empty string.

---

Other binding technologies have configurable null/DBNull behavior. Windows forms has [Binding.NullValue](https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.binding.nullvalue?view=windowsdesktop-9.0) and [Binding.DataSourceNullValue](https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.binding.datasourcenullvalue?view=windowsdesktop-9.0). Having a similar tranformation/interpretation may also be an option.

Otherwise, given that HTML select value always converts a set null value to an empty string value, null must always be a non-match.

Or it must be defined to always be interpreted as an empty string value which must also lead an empty value option being selected or non-match if none exists without any option selected or displayed.

Was was mentioned initially, *seeing* something different than what is the *bound value* is very misleading and user-interaction error-prone.

It seems like implementing a never-match-null through a magic value seems reasonably simple/straight-forward and would resolve bind value display mismatches.

Null as a selectable option was not covered before without custom code (and a defined transformation), and would not be covered after in this simple solution either.

---

<details><summary>HTML test code</summary>

```html
a <!-- default value "va" -->
<select id="a">
<option value="va">A</option>
<option value="vb">B</option>
</select>

a2 <!-- default value "va" -->
<select id="a2">
<option value="va">A</option>
<option value="vb">B</option>
</select>
<script>document.getElementById('a2').value = 'vb'</script>

a3 <!-- default value "va" -->
<select id="a3">
<option value="va">A</option>
<option value="vb">B</option>
</select>
<script>document.getElementById('a3').value = 'INVALID'</script>

b <!-- default value "" -->
<select id="b" autocomplete="off">
<option></option>
<option value="va">A</option>
<option value="vb">B</option>
</select>

c <!-- default value "" -->
<select id="c" autocomplete="off">
<option value=""></option>
<option value="va">A</option>
<option value="vb">B</option>
</select>

d <!-- default value "va" -->
<select id="d" autocomplete="off">
<option value="va">A</option>
<option></option>
<option value="vb">B</option>
</select>

e <!-- default value "" -->
<select id="e" autocomplete="off">
<option value="va">A</option>
<option selected></option>
<option value="vb">B</option>
</select>
```

</details>
